src/extras/primitives/registerPrimitive.js

```diff
@@ -75,6 +69,33 @@ module.exports = function registerPrimitive (name, definition) {
         }
       },
 
+      applyDefaultComponents: {
+        value: function () {
+          var self = this;
+          var defaultData = cloneObject(this.defaultAttributes);
+
+          // Apply default components.
+          Object.keys(defaultData).forEach(function applyDefault (componentName) {
+            var componentData = defaultData[componentName];
+
+            // Set component properties individually to not overwrite user-defined components.
+            if (componentData instanceof Object && Object.keys(componentData).length) {
+              Object.keys(componentData).forEach(function setProperty (propName) {
+                // Check if component property already defined.
+                var definedData = self.getAttribute(componentName);
+                if (definedData && definedData[propName] !== undefined) { return; }
+
+                self.setAttribute(componentName, propName, componentData[propName]);
+              });
+              return;
```

Comment:
I'm only returning here to mimic an `if/else`. Not a big fan of that style since it's less clear on the intent, but dmarcos likes that.